### ISSUE TITLE
Feature [High] - Basic and Advanced SBO Screens

### ISSUE THREAD

**Issue by JeneralPain** [rollerderby/scoreboard#459]
Following on from the significant community feedback around the operator panel being too advanced when you only need "basic scoring" features, we have discussed the creation of...

- Basic SBO Panel (basically like the old 3.9.5) with no stats recording beyond the score
- Advanced SBO Panel (same as 4.x) with full stats recording

The basic panel wont have:
- Score tracking pages/panels
- Won't record NI/Trip data
- Will just display the score(s) in the most basic sense

The advanced panel:
Same as the 4.x panel is now.

The community has provided quite loud and honest feedback around this and it's vital we do this sooner rather than later. because many have said they're going to go back to 3.9.x for simplicity.

**Comment by frank-weinberg** [comment]
I think this is a bad idea:

* The problems were primarily reported by people who had gained experience with 3.9.x and then switched over to 4.x (and as it would seem often without bothering to check the documentation or trying it at home/a scrimmage first). On the other hand in the same thread there are reports that new people who start directly with 4.x have a lot less trouble getting used to the new interface. If we now create a 3.9.5-like interface and present it as what beginners should use first, we would be setting up all new SBOs for having to go through the apparently difficult transition from the basic to the full interface.

* Having these two interfaces permanently is bound to set up situations where an SBO who only knows the basic interface is rostered for an event where CRG stats generation is supposed to be used. This won't end well.

* Not properly recording NI/Trip data (you can't not record it at all without serious changes to the backend) will affect the public display as that indicates NI. So we would need adjustments to that screen as well and would need to make sure they are aligned with the choice of operator screen.

* One of the principles @brian-brazil and me are in perfect agreement about is that we don't want to maintain multiple screens for the same functionality. This proposal runs against that principle. And this proposed "basic" screen would be the opposite of basic in terms of code compllexity as it would have to address a ton of edge cases in order to avoid unwanted side effects. (One example: User tries to reduce the jam number - you have to delete a jam. But deleting a jam deletes the points in the jam. So you have to either find a jam without points (What do you do when there is none?) or move the points elsewhere (What if it's the only jam?). And any future change in CRG would have to be checked for it's interactions with this screen. If you find someone who is masochistic enough to create and maintain this as a custom screen, I won't stop them. But I flat out refuse to take any responsibility for a behemoth like this by including it in the main distribution.


Instead I am working on the following course of action:

- Improve the documentation on the Wiki. (Mostly done for the moment - I want to add a "Less common situations" like I already have for LT to the SBO page as well that should cover at least all the things that people asked about on Facebook or here.)
- Put together a pdf manual consisting of the relevant wiki pages and include it with the release zip. Put a link to the pdf on the start page.
- Maybe put a note on the release page that you should only use 4.x in a public game if your SBO has had time to familiarize themselves with the new interface (or has someone on hand who can teach them as they go).

> because many have said they're going to go back to 3.9.x for simplicity.

If all they need is 3.9.5 and they don't want/have the time to learn the new interface then they should use 3.9.5.

**Comment by riffreff** [comment]
I totally agree that many of the "issues" users have with the new controls is because they just fired up 4.0.x and because it looks very similar to the previous version expected everything to work the same way. I mean, how could you use it in a tournament without testing the most basic functions like correcting score and jam numbers? That is the same issue, as one user suggested, that the new version is not in sync with last years WFTDA tournament instructions about ending TOs. Which we know is because of that one overlay not being able to go from TO into Lineup.

However...

I can imagine it being very useful to have a simplified interface without stats (displayed) and the "old" way of correcting TO#, Jam# asf.
I would go as far as not including Jammers (and Pivots) in that interface. Basically, you just Start Stop TO, and Score. That's it. No Jammer names or star passes.

I would suggest that to the users, and if they then say: No that's too basic, then I would refer them to 3.9.5, because then we know that these users just wanted the impossible "better scoreboard software but no changes".

**Comment by frank-weinberg** [comment]
If you believe that such a "simplified" screen is worth the substantial effort of creating it, go ahead and do that. The custom screens folder has been created exactly for things like that.

If that custom screen turns out to be a roaring success and someone proposes inclusion along with a commitment to maintaining it here, I wouldn't even block that just out of principle. (In that situation the issues from my first two bullet points would presumably have manifested anyways, so trying to avoid them through not creating the screen becomes a moot point.) But if the commitment to maintain drops and isn't picked up by someone else, the screen probably won't stay long.

**Comment by AdamSmasherDerby** [comment]
@riffreff sadly, it's not as simple as one might hope, because there's going to have to be some torturing of the backend data structure to allow for things like "score -1" buttons to ignore trip boundaries.  It will CERTAINLY be impossible to switch from "lite" to "full stats" midgame.

That said, this is something that there is MASSIVE demand for.  We cannot simply make the software what we want it to be while ignoring user needs.  And a big segment of the user base needs this to work more simply.

Directing them back to 3.9.5 is not a great option, and we're all aware of why, but that IS what's going to happen.  Anecdotally, I already know of a number of experienced NSOs who have reverted.

I'm going to start working on this screen, but we MUST make sure we continue to remember that developers are here to serve the users, not the other way around.

**Comment by frank-weinberg** [comment]
> developers are here to serve the users not the other way around.

False dichotomy. And developing software where others get to decide what and how I implement is what I do as a paid job, not as a hobby. Users can ask or try to convince me to develop something but they don't get to demand. (Just like I don't get to demand that they use what I write.)

And FTR: I'm going to stop reading the public Facebook group at least for a while. Seeing a constant stream of people expressing disappointment with my work and not having a chance to explain myself and my choices there is just too mentally exhausting. (And to be clear: I'm not saying that any individual post is over the line. But the sum of them very much *feels* like a shit storm. Which is why I used those words in an earlier commment. I realize now that this was an overreaction and apologize for the language. I have edited that paragraph.)

**Comment by official-sounding** [comment]
> because there's going to have to be some torturing of the backend data structure to allow for things like "score -1" buttons to ignore trip boundaries.

From an architecture perspective, I think it would make more sense to turn off trip-auto-increment and just dump all the points into trip 2 (trip 2 also "solves" any NI issues).  Since you're not using the trip data anyways, no need to try to keep things in trips behind the scenes.

I was under the impression that auto-trip-increment was a user-facing setting somewhere.  Is that a "ruleset" or "scoreboard" setting (or neither, and I was totally mistaken?) and done from the backend.  Assuming its a scoreboard setting, you could turn it off, but then would potentially need some mechanism to reset the preference so you don't break the main scoreboard. That moves it out of the realm of just a custom screen, though.

As an aside, I'd wish that we could move the needle and allow for more "code complexity" in the service of trying to make user screens more intuitive and user-friendly.  I can certainly appreciate being suspect, as often "less user friendly" just means "different", but it does feel like there is some legitimate feedback asking for ways to improve the interface.  It is extremely demotivating to try to contribute to this project when it can feel like such an uphill battle to justify the changes philosophically, let alone technically.

**Comment by riffreff** [comment]
>  But the sum of them very much feels like a shit storm.

It's not a shit storm, it is "waaaaaaaah I hate change!" We have to take it all with a right amount of salt. I don't have much sympathy for the ones who just start to look at CRG the day before their big game or tournament and then are surprised that they ran into issues with the UI.
But I can see the benefits of a simplified version. I totally get your point of view and I agree, that this is an open source project and everyone is free to get involved. I think we don't have to do anything atm. We should give 4.0.X time and people to get used to it. 3.9.5. is still an alternative for "haters"

**Comment by frank-weinberg** [comment]
>  turn off trip-auto-increment

There is currently no way to do that. #450 introduces the possibility to cancel individual instances via "Trip -1".

> I'd wish that we could move the needle and allow for more "code complexity" in the service of [...] It is extremely demotivating to try to contribute to this project when it can feel like such an uphill battle to justify the changes philosophically

It looks like that move may have happened two weeks ago, when Brain declared he would be unwilling to contribute to a project where merges can happen despite objections by a maintainer and I declared that I would be unwilling to contribute to a project where any single person can veto any change. Brain hasn't had any visible interaction with this repo since and has drastically reduced his activity in the Facebook group. If this continues a bit longer, I will assume that he is at least taking a break as maintainer and act accordingly.

In that case I would work with the following list of valid reasons for not merging pull requests right away:

- It breaks a currently working use case. This can often (but not always) be resolved by making the feature optional. Where options don't work, there may be other solutions.
- There is nobody to maintain this. This can obviously be resolved by providing someone who does. Breaking this feature is then not a reason in itself to block a PR down the road. But making the maintainer's job impossible or unneccessarily hard is.
- Code quality / UI style issues. The objection should make clear what the problems are and what is expected as a fix.
- Matters of taste/preference as well as scope extensions can be brought up as suggestions but the person who puts in the work to implement a feature should has the last say on these matters.
  - On pull requests that seek to override a previous choice on a matter of preference (without the change adding functionality) the original feature author has a veto unless the new choice is backed by a substantial preference in the user base.

I would also expect reviews to state all objections right away or at least indicate areas in which objections may come up later (e.g. "I didn't check the code in detail yet."). 

I am willing to discuss these guidelines, but if anybody wants to do so, I'd request we move that to its own ticket.

>> feels like a shit storm

> It's not 

That's why I emphasized the "feels".

**Comment by AdamSmasherDerby** [comment]
> > But the sum of them very much feels like a shit storm.
> 
> It's not a shit storm, it is "waaaaaaaah I hate change!" We have to take it all with a right amount of salt. I don't have much sympathy for the ones who just start to look at CRG the day before their big game or tournament and then are surprised that they ran into issues with the UI.

I think we need to appreciate that the MAJORITY of games using this software are not only not at tournaments, they aren't even sanctioned.  This software is the single indispensable tool that basically every public WFTDA / MRDA / JRDA rules game worldwide _depends_ on.  

Coding as though our only audience was experienced, practiced SBOs producing stats for a sanctioned game is ignoring not only an important use case of our software, I would argue it's probably even the _dominant_ use case.

It's great that we're extending the reach of CRG's features.  I'm stoked that we have a Rinxter killer.

But pretty much all of derby depends on this software to work. We can't act as though it's fine to build it exactly the way WE want it, because we are not the only people using it.  There has to be some empathy for the dozens (hundreds?) of games that happen every weekend that just need it to _work_ and work easily, even when run by inexperienced operators.

**Comment by frank-weinberg** [comment]
I tried to make CRG with stats as easy to operate as I possibly could. And in the >11 months since I first gave a version to other people there have been exactly TWO (2) suggestions how the SBO interface could be made easier to use that did not involve completely breaking the stats. I take this as an indication that there is not all that much potential for improvement left. So if this long term is too complicated for a substantial part of derby then we will most likely need both CRG with stats and CRG without stats. And if the creation of CRG with stats had made CRG without stats unavailable, you would be right in demanding we do something about that. But it hasn't. 3.9.5 is still available and can be used as easily as before.

So apparently you seem to believe that by including some features in CRG with stats that CRG without stats could also use, we are somehow obligated to create a CRG without stats that includes these features. But I don't see why that should be the case.

***
And if I look at this feature request as one for creating an improved CRG without stats then I can see 4 ways of going at that:

1. Adding more complexity on top of CRG with stats in order to hide the complexity caused by the stats, as suggested here.
2. Taking old CRG without stats (3.9.5 or some dev revision between that release and adding stats support) and porting back the features people want. 
3. Removing features from CRG with stats that CRG without stats doesn't need.
4. Starting from scratch.

Of those 4. is probably the most of work by a lot. 1. also has a substantial chance of failing by virtue of requiring a lot of complexity. And this will only get worse over time. Of the other two, 3. would probably be more work initially but it would be easier to port features between the two CRG versions once the initial work is done.

(And as I think about it, the backend part of 3. is something I could potentially enjoy doing.)

**Comment by AdamSmasherDerby** [comment]
I don't plan on doing anything else with the backend or making any more pull requests to the repo unless another truly game breaking issue arises. (On the order of defaulting to an impossible game state.) It's not worth the philosophical arguments.

I'm pretty sure I can implement a "lite" scoreboard screen purely as a custom frontend page, and that's my plan.  It's not hard to figure out how to work around the backend issues - for example, the "NI" indication is fixed by simply advancing to the second scoring trip immediately on jam start.  That means it doesn't work at all, but that's the point - if you want good stats, run the full version, and if you want simplicity, you don't get stats.

If anyone else wants to take a different approach to the same problem - power to them!  There should be room in the world for more than 2 or 3 developers working on different aspects of this project.
